File: src/utils/summarize.ts
/**
 * Tool Summarizer
 * Maps tool names and inputs to human-readable one-liners.
 */

/**
 * Show last 2 path components with .../ prefix.
 * E.g. `/opt/project/src/utils/config.ts` -> `.../utils/config.ts`
 */
export function shortPath(filePath: string): string {
  const parts = filePath.split('/').filter(Boolean);
  if (parts.length <= 2) return filePath;
  return `.../${parts.slice(-2).join('/')}`;
}

/**
 * Show the hostname (or first 40 chars) of a URL.
 */
function shortUrl(url: string): string {
  try {
    const parsed = new URL(url);
    return parsed.hostname;
  } catch {
    return url.slice(0, 40);
  }
}

/**
 * Truncate a string to a given length, appending "..." if truncated.
 */
function truncate(str: string, maxLen: number): string {
  if (str.length <= maxLen) return str;
  return str.slice(0, maxLen) + '...';
}

// Commands that are trivial prefixes in chained commands
const TRIVIAL_COMMANDS = new Set([
  'cd', 'export', 'echo', 'sleep', 'set', 'source', 'true', 'false', ':'
]);

// Wrapper commands that should be stripped to get the real command
const WRAPPER_COMMANDS = new Set([
  'sudo', 'nohup', 'timeout', 'env', 'nice'
]);

/**
 * For chained commands (&&, ;, ||), skip trivial prefixes
 * and return the first meaningful command.
 * Pipes are NOT split -- they are part of the same command.
 */
export function findMeaningfulCommand(command: string): string {
  // Split on &&, ;, || but not | (single pipe)
  // Use a regex that matches && or || or ; but not a lone |
  const segments = command.split(/\s*(?:&&|;|\|\|)\s*/);

  for (const segment of segments) {
    const trimmed = segment.trim();
    if (!trimmed) continue;
    const firstWord = trimmed.split(/\s+/)[0];
    if (!TRIVIAL_COMMANDS.has(firstWord)) {
      return trimmed;
    }
  }

  // If all segments are trivial, return the last one (or original)
  const lastSegment = segments[segments.length - 1]?.trim();
  return lastSegment || command.trim();
}

/**
 * Strip wrapper commands (sudo, nohup, etc.) and return the inner command.
 */
function stripWrappers(command: string): string {
  const parts = command.trim().split(/\s+/);
  let i = 0;
  while (i < parts.length && WRAPPER_COMMANDS.has(parts[i])) {
    i++;
    // For 'timeout' and 'env', skip one extra argument
    // timeout has a duration arg: `timeout 30 cmd`
    // env may have VAR=val pairs
    if (i < parts.length) {
      const prev = parts[i - 1];
      if (prev === 'timeout' && i < parts.length && /^[\d.]+[smhd]?$/.test(parts[i])) {
        i++;
      }
    }
  }
  if (i >= parts.length) return command.trim();
  return parts.slice(i).join(' ');
}

/**
 * Summarize a Bash command to a human-readable description.
 */
function summarizeBashCommand(rawCommand: string): string {
  // First find the meaningful command in chains
  const meaningful = findMeaningfulCommand(rawCommand);
  // Then strip wrappers
  const command = stripWrappers(meaningful);
  const parts = command.split(/\s+/);
  const first = parts[0];

  // cargo commands
  if (first === 'cargo') {
    const sub = parts[1];
    if (sub === 'build') {
      const isRelease = parts.includes('--release');
      return isRelease ? 'Building project (release)' : 'Building project';
    }
    if (sub === 'test') return 'Running tests';
    if (sub === 'clippy') return 'Running linter';
    if (sub === 'fmt') return 'Formatting code';
    if (sub === 'run') return 'Running project';
    if (sub === 'add') return 'Adding dependency';
    if (sub === 'install') return 'Installing tool';
    if (sub === 'clean') return 'Cleaning build';
    if (sub === 'doc') return 'Generating docs';
    if (sub === 'publish') return 'Publishing crate';
    if (sub === 'bench') return 'Running benchmarks';
    if (sub === 'check') return 'Type checking';
    return `Running \`cargo ${sub || ''}\``;
  }

  // git commands
  if (first === 'git') {
    const sub = parts[1];
    if (sub === 'clone') return 'Cloning repository';
    if (sub === 'commit') return 'Committing changes';
    if (sub === 'push') return 'Pushing to remote';
    if (sub === 'pull') return 'Pulling from remote';
    if (sub === 'checkout' || sub === 'switch') return 'Switching branch';
    if (sub === 'merge') return 'Merging branches';
    if (sub === 'rebase') return 'Rebasing';
    if (sub === 'stash') return 'Stashing changes';
    if (sub === 'diff') return 'Viewing diff';
    if (sub === 'log') return 'Viewing history';
    if (sub === 'status') return 'Checking status';
    if (sub === 'branch') return 'Managing branches';
    if (sub === 'tag') return 'Managing tags';
    if (sub === 'fetch') return 'Fetching remote';
    if (sub === 'reset') return 'Resetting changes';
    return `Running \`git ${sub || ''}\``;
  }

  // npm commands
  if (first === 'npm') {
    const sub = parts[1];
    if (sub === 'install' || sub === 'ci' || sub === 'i') return 'Installing dependencies';
    if (sub === 'run') {
      const script = parts[2];
      if (script === 'build') return 'Building project';
      if (script === 'test') return 'Running tests';
      if (script === 'lint') return 'Running linter';
      return `Running npm script: ${script || 'unknown'}`;
    }
    if (sub === 'test' || sub === 't') return 'Running tests';
    if (sub === 'publish') return 'Publishing package';
    return `Running \`npm ${sub || ''}\``;
  }

  // npx
  if (first === 'npx') {
    const pkg = parts[1] || 'unknown';
    return `Running npx: ${pkg}`;
  }

  // yarn
  if (first === 'yarn') {
    const sub = parts[1];
    if (!sub || sub === 'install') return 'Installing dependencies';
    if (sub === 'build') return 'Building project';
    if (sub === 'test') return 'Running tests';
    if (sub === 'lint') return 'Running linter';
    if (sub === 'add') return 'Installing dependencies';
    if (sub === 'publish') return 'Publishing package';
    return `Running \`yarn ${sub}\``;
  }

  // pnpm
  if (first === 'pnpm') {
    const sub = parts[1];
    if (sub === 'install' || sub === 'i') return 'Installing dependencies';
    if (sub === 'run') {
      const script = parts[2];
      if (script === 'build') return 'Building project';
      if (script === 'test') return 'Running tests';
      if (script === 'lint') return 'Running linter';
      return `Running pnpm script: ${script || 'unknown'}`;
    }
    if (sub === 'test') return 'Running tests';
    if (sub === 'add') return 'Installing dependencies';
    if (sub === 'publish') return 'Publishing package';
    return `Running \`pnpm ${sub || ''}\``;
  }

  // bun
  if (first === 'bun') {
    const sub = parts[1];
    if (sub === 'install' || sub === 'i') return 'Installing dependencies';
    if (sub === 'run') {
      const script = parts[2];
      if (script === 'build') return 'Building project';
      if (script === 'test') return 'Running tests';
      if (script === 'lint') return 'Running linter';
      return `Running bun script: ${script || 'unknown'}`;
    }
    if (sub === 'test') return 'Running tests';
    if (sub === 'add') return 'Installing dependencies';
    if (sub === 'publish') return 'Publishing package';
    return `Running \`bun ${sub || ''}\``;
  }

  // Python
  if (first === 'pip' || first === 'pip3') {
    if (parts[1] === 'install') return 'Installing Python dependencies';
    return `Running \`${first} ${parts[1] || ''}\``;
  }
  if (first === 'pytest' || (first === 'python' && parts[1] === '-m' && parts[2] === 'pytest')) {
    return 'Running Python tests';
  }
  if (first === 'python' || first === 'python3') {
    if (parts[1] === '-m' && parts[2] === 'pytest') return 'Running Python tests';
    return `Running \`${first}\``;
  }

  // Docker
  if (first === 'docker') {
    if (parts[1] === 'build') return 'Building Docker image';
    if (parts[1] === 'run') return 'Running container';
    return `Running \`docker ${parts[1] || ''}\``;
  }
  if (first === 'docker-compose' || first === 'docker compose') {
    if (parts[1] === 'up') return 'Starting containers';
    return `Running \`docker-compose ${parts[1] || ''}\``;
  }

  // make
  if (first === 'make') {
    const target = parts[1];
    return target ? `Running make: ${target}` : 'Building with make';
  }

  // TypeScript / JS tools
  if (first === 'tsc') return 'Type checking';
  if (first === 'vitest') return 'Running tests';
  if (first === 'eslint') return 'Running linter';

  // Network tools
  if (first === 'curl') return 'Fetching URL';
  if (first === 'wget') return 'Downloading file';
  if (first === 'ssh') return 'Connecting via SSH';
  if (first === 'scp') return 'Copying files via SSH';

  // File tools
  if (first === 'tar') return 'Archiving/extracting';
  if (first === 'chmod') return 'Changing permissions';
  if (first === 'chown') return 'Changing ownership';
  if (first === 'mkdir') return 'Creating directory';
  if (first === 'rm') return 'Removing files';
  if (first === 'cp') return 'Copying files';
  if (first === 'mv') return 'Moving files';

  // Search
  if (first === 'grep' || first === 'rg') return 'Searching files';
  if (first === 'find') return 'Finding files';

  // Infrastructure
  if (first === 'kubectl') return 'Managing Kubernetes';
  if (first === 'terraform') return 'Managing infrastructure';

  // Go
  if (first === 'go') {
    const sub = parts[1];
    if (sub === 'build') return 'Building Go project';
    if (sub === 'test') return 'Running Go tests';
    if (sub === 'run') return 'Running Go project';
    return `Running \`go ${sub || ''}\``;
  }

  // Rust
  if (first === 'rustc') return 'Compiling Rust';

  // Fallback
  return `Running \`${first}\``;
}

/**
 * Maps tool name + input to a human-readable one-liner.
 */
export function summarizeToolAction(tool: string, input: Record<string, unknown>): string {
  if (tool === 'Bash') {
    const command = input.command as string | undefined;
    if (command) {
      return summarizeBashCommand(command);
    }
    return 'Running command';
  }

  if (tool === 'Read') {
    const fp = input.file_path as string | undefined;
    return fp ? `Reading ${shortPath(fp)}` : 'Reading file';
  }

  if (tool === 'Write') {
    const fp = input.file_path as string | undefined;
    return fp ? `Writing ${shortPath(fp)}` : 'Writing file';
  }

  if (tool === 'Edit' || tool === 'MultiEdit') {
    const fp = input.file_path as string | undefined;
    return fp ? `Editing ${shortPath(fp)}` : 'Editing file';
  }

  if (tool === 'Grep') {
    const pattern = input.pattern as string | undefined;
    return pattern ? `Searching for '${truncate(pattern, 30)}'` : 'Searching files';
  }

  if (tool === 'Glob') {
    const pattern = input.pattern as string | undefined;
    return pattern ? `Finding files: ${pattern}` : 'Finding files';
  }

  if (tool === 'Task') return 'Running task';

  if (tool === 'WebSearch') {
    const query = input.query as string | undefined;
    return query ? `Searching: ${truncate(query, 40)}` : 'Searching the web';
  }

  if (tool === 'WebFetch') {
    const url = input.url as string | undefined;
    return url ? `Fetching ${shortUrl(url)}` : 'Fetching URL';
  }

  if (tool === 'TodoWrite') return 'Updating tasks';
  if (tool === 'TodoRead') return 'Reading tasks';
  if (tool === 'AskUserQuestion') return 'Asking user a question';
  if (tool === 'NotebookEdit') return 'Editing notebook';

  return `Using ${tool}`;
}

/**
 * Detect error patterns in tool output and return a summary.
 */
export function summarizeToolResult(_tool: string, output: string): string {
  if (!output) return 'Completed (no output)';

  const lines = output.split('\n');

  // Check for error patterns
  for (const line of lines) {
    // Rust compiler error
    if (line.includes('error[E')) {
      return `Failed: ${truncate(line.trim(), 60)}`;
    }
  }

  for (const line of lines) {
    // Generic Error:
    if (/\bError:/i.test(line)) {
      return `Failed: ${truncate(line.trim(), 60)}`;
    }
  }

  // FAILED pattern (typically test output)
  if (output.includes('FAILED')) {
    return 'Tests failed';
  }

  // Panic
  for (const line of lines) {
    if (line.includes('panic!') || line.includes('panicked at')) {
      return `Panicked: ${truncate(line.trim(), 60)}`;
    }
  }

  // npm error
  if (output.includes('npm ERR!')) {
    return 'npm error';
  }

  // Default
  const lineCount = lines.length;
  return `Completed (${lineCount} lines of output)`;
}

File: rust-crates/ctm/src/summarize.rs
//! Tool summarizer — maps tool names + inputs to human-readable one-liners.
//!
//! Ported from `summarize.ts` with all 30+ Bash command patterns.

use std::collections::HashSet;
use std::sync::LazyLock;

use crate::formatting::{short_path, truncate};

// ------------------------------------------------------------------ helpers

/// Extract hostname from a URL, falling back to first 40 chars.
fn short_url(url: &str) -> String {
    // Minimal URL parse — look for "://" and take the host part.
    if let Some(rest) = url
        .strip_prefix("https://")
        .or_else(|| url.strip_prefix("http://"))
    {
        if let Some(host) = rest.split('/').next() {
            return host.to_string();
        }
    }
    truncate(url, 40)
}

// -------------------------------------------------------- trivial / wrapper

static TRIVIAL_COMMANDS: LazyLock<HashSet<&'static str>> = LazyLock::new(|| {
    [
        "cd", "export", "echo", "sleep", "set", "source", "true", "false", ":",
    ]
    .into_iter()
    .collect()
});

static WRAPPER_COMMANDS: LazyLock<HashSet<&'static str>> = LazyLock::new(|| {
    ["sudo", "nohup", "timeout", "env", "nice"]
        .into_iter()
        .collect()
});

/// For chained commands (`&&`, `;`, `||`), skip trivial prefixes.
/// Pipes are NOT split — they are part of the same command.
pub fn find_meaningful_command(command: &str) -> String {
    // Split on &&, ;, || but NOT single |
    let segments: Vec<&str> = command
        .split("&&")
        .flat_map(|s| s.split(';'))
        .flat_map(|s| split_on_double_pipe(s))
        .collect();

    for segment in &segments {
        let trimmed = segment.trim();
        if trimmed.is_empty() {
            continue;
        }
        let first_word = trimmed.split_whitespace().next().unwrap_or("");
        if !TRIVIAL_COMMANDS.contains(first_word) {
            return trimmed.to_string();
        }
    }

    // All trivial — return last segment.
    segments
        .last()
        .map(|s| s.trim().to_string())
        .unwrap_or_else(|| command.trim().to_string())
}

/// Split on `||` only, not single `|`.
fn split_on_double_pipe(s: &str) -> Vec<&str> {
    let mut results = Vec::new();
    let mut start = 0;
    let bytes = s.as_bytes();
    let mut i = 0;

    while i < bytes.len() {
        if i + 1 < bytes.len() && bytes[i] == b'|' && bytes[i + 1] == b'|' {
            results.push(&s[start..i]);
            start = i + 2;
            i += 2;
        } else {
            i += 1;
        }
    }
    results.push(&s[start..]);
    results
}

/// Strip wrapper commands (sudo, nohup, etc.) and return the inner command.
fn strip_wrappers(command: &str) -> String {
    let cmd = command.trim();
    let parts: Vec<&str> = cmd.split_whitespace().collect();
    let mut i = 0;
    while i < parts.len() && WRAPPER_COMMANDS.contains(parts[i]) {
        let prev = parts[i];
        i += 1;
        // `timeout` has a duration arg: `timeout 30 cmd`
        if prev == "timeout" && i < parts.len() {
            // Check if next looks like a duration (digits with optional suffix)
            let next = parts[i];
            if next.chars().next().is_some_and(|c| c.is_ascii_digit()) {
                i += 1;
            }
        }
    }
    if i >= parts.len() {
        return command.trim().to_string();
    }
    parts[i..].join(" ")
}

// -------------------------------------------------------- bash summarizer

fn summarize_bash_command(raw_command: &str) -> String {
    let meaningful = find_meaningful_command(raw_command);
    let command = strip_wrappers(&meaningful);
    let parts: Vec<&str> = command.split_whitespace().collect();
    let first = parts.first().copied().unwrap_or("");

    match first {
        "cargo" => {
            let sub = parts.get(1).copied().unwrap_or("");
            match sub {
                "build" => {
                    if parts.contains(&"--release") {
                        "Building project (release)".to_string()
                    } else {
                        "Building project".to_string()
                    }
                }
                "test" => "Running tests".to_string(),
                "clippy" => "Running linter".to_string(),
                "fmt" => "Formatting code".to_string(),
                "run" => "Running project".to_string(),
                "add" => "Adding dependency".to_string(),
                "install" => "Installing tool".to_string(),
                "clean" => "Cleaning build".to_string(),
                "doc" => "Generating docs".to_string(),
                "publish" => "Publishing crate".to_string(),
                "bench" => "Running benchmarks".to_string(),
                "check" => "Type checking".to_string(),
                _ => format!("Running `cargo {sub}`"),
            }
        }

        "git" => {
            let sub = parts.get(1).copied().unwrap_or("");
            match sub {
                "clone" => "Cloning repository".to_string(),
                "commit" => "Committing changes".to_string(),
                "push" => "Pushing to remote".to_string(),
                "pull" => "Pulling from remote".to_string(),
                "checkout" | "switch" => "Switching branch".to_string(),
                "merge" => "Merging branches".to_string(),
                "rebase" => "Rebasing".to_string(),
                "stash" => "Stashing changes".to_string(),
                "diff" => "Viewing diff".to_string(),
                "log" => "Viewing history".to_string(),
                "status" => "Checking status".to_string(),
                "branch" => "Managing branches".to_string(),
                "tag" => "Managing tags".to_string(),
                "fetch" => "Fetching remote".to_string(),
                "reset" => "Resetting changes".to_string(),
                _ => format!("Running `git {sub}`"),
            }
        }

        "npm" => {
            let sub = parts.get(1).copied().unwrap_or("");
            match sub {
                "install" | "ci" | "i" => "Installing dependencies".to_string(),
                "run" => {
                    let script = parts.get(2).copied().unwrap_or("unknown");
                    match script {
                        "build" => "Building project".to_string(),
                        "test" => "Running tests".to_string(),
                        "lint" => "Running linter".to_string(),
                        _ => format!("Running npm script: {script}"),
                    }
                }
                "test" | "t" => "Running tests".to_string(),
                "publish" => "Publishing package".to_string(),
                _ => format!("Running `npm {sub}`"),
            }
        }

        "npx" => {
            let pkg = parts.get(1).copied().unwrap_or("unknown");
            format!("Running npx: {pkg}")
        }

        "yarn" => {
            let sub = parts.get(1).copied();
            match sub {
                None | Some("install") => "Installing dependencies".to_string(),
                Some("build") => "Building project".to_string(),
                Some("test") => "Running tests".to_string(),
                Some("lint") => "Running linter".to_string(),
                Some("add") => "Installing dependencies".to_string(),
                Some("publish") => "Publishing package".to_string(),
                Some(s) => format!("Running `yarn {s}`"),
            }
        }

        "pnpm" => {
            let sub = parts.get(1).copied().unwrap_or("");
            match sub {
                "install" | "i" => "Installing dependencies".to_string(),
                "run" => {
                    let script = parts.get(2).copied().unwrap_or("unknown");
                    match script {
                        "build" => "Building project".to_string(),
                        "test" => "Running tests".to_string(),
                        "lint" => "Running linter".to_string(),
                        _ => format!("Running pnpm script: {script}"),
                    }
                }
                "test" => "Running tests".to_string(),
                "add" => "Installing dependencies".to_string(),
                "publish" => "Publishing package".to_string(),
                _ => format!("Running `pnpm {sub}`"),
            }
        }

        "bun" => {
            let sub = parts.get(1).copied().unwrap_or("");
            match sub {
                "install" | "i" => "Installing dependencies".to_string(),
                "run" => {
                    let script = parts.get(2).copied().unwrap_or("unknown");
                    match script {
                        "build" => "Building project".to_string(),
                        "test" => "Running tests".to_string(),
                        "lint" => "Running linter".to_string(),
                        _ => format!("Running bun script: {script}"),
                    }
                }
                "test" => "Running tests".to_string(),
                "add" => "Installing dependencies".to_string(),
                "publish" => "Publishing package".to_string(),
                _ => format!("Running `bun {sub}`"),
            }
        }

        "pip" | "pip3" => {
            if parts.get(1).copied() == Some("install") {
                "Installing Python dependencies".to_string()
            } else {
                format!(
                    "Running `{} {}`",
                    first,
                    parts.get(1).copied().unwrap_or("")
                )
            }
        }

        "pytest" => "Running Python tests".to_string(),

        "python" | "python3" => {
            if parts.get(1).copied() == Some("-m") && parts.get(2).copied() == Some("pytest") {
                "Running Python tests".to_string()
            } else {
                format!("Running `{first}`")
            }
        }

        "docker" => {
            let sub = parts.get(1).copied().unwrap_or("");
            // Handle "docker compose" (two-word modern syntax)
            if sub == "compose" {
                let compose_sub = parts.get(2).copied().unwrap_or("");
                return if compose_sub == "up" {
                    "Starting containers".to_string()
                } else {
                    format!("Running `docker compose {compose_sub}`")
                };
            }
            match sub {
                "build" => "Building Docker image".to_string(),
                "run" => "Running container".to_string(),
                _ => format!("Running `docker {sub}`"),
            }
        }

        "docker-compose" => {
            let sub = parts.get(1).copied().unwrap_or("");
            if sub == "up" {
                "Starting containers".to_string()
            } else {
                format!("Running `docker-compose {sub}`")
            }
        }

        "make" => {
            let target = parts.get(1).copied();
            match target {
                Some(t) => format!("Running make: {t}"),
                None => "Building with make".to_string(),
            }
        }

        "tsc" => "Type checking".to_string(),
        "vitest" => "Running tests".to_string(),
        "eslint" => "Running linter".to_string(),

        "curl" => "Fetching URL".to_string(),
        "wget" => "Downloading file".to_string(),
        "ssh" => "Connecting via SSH".to_string(),
        "scp" => "Copying files via SSH".to_string(),

        "tar" => "Archiving/extracting".to_string(),
        "chmod" => "Changing permissions".to_string(),
        "chown" => "Changing ownership".to_string(),
        "mkdir" => "Creating directory".to_string(),
        "rm" => "Removing files".to_string(),
        "cp" => "Copying files".to_string(),
        "mv" => "Moving files".to_string(),

        "grep" | "rg" => "Searching files".to_string(),
        "find" => "Finding files".to_string(),

        "kubectl" => "Managing Kubernetes".to_string(),
        "terraform" => "Managing infrastructure".to_string(),

        "go" => {
            let sub = parts.get(1).copied().unwrap_or("");
            match sub {
                "build" => "Building Go project".to_string(),
                "test" => "Running Go tests".to_string(),
                "run" => "Running Go project".to_string(),
                _ => format!("Running `go {sub}`"),
            }
        }

        "rustc" => "Compiling Rust".to_string(),

        _ => format!("Running `{first}`"),
    }
}

// -------------------------------------------------------- public API

/// Map tool name + input to a human-readable one-liner.
pub fn summarize_tool_action(tool: &str, input: &serde_json::Value) -> String {
    match tool {
        "Bash" => {
            let command = input.get("command").and_then(|v| v.as_str()).unwrap_or("");
            if command.is_empty() {
                "Running command".to_string()
            } else {
                summarize_bash_command(command)
            }
        }
        "Read" => {
            let fp = input.get("file_path").and_then(|v| v.as_str());
            match fp {
                Some(p) => format!("Reading {}", short_path(p)),
                None => "Reading file".to_string(),
            }
        }
        "Write" => {
            let fp = input.get("file_path").and_then(|v| v.as_str());
            match fp {
                Some(p) => format!("Writing {}", short_path(p)),
                None => "Writing file".to_string(),
            }
        }
        "Edit" | "MultiEdit" => {
            let fp = input.get("file_path").and_then(|v| v.as_str());
            match fp {
                Some(p) => format!("Editing {}", short_path(p)),
                None => "Editing file".to_string(),
            }
        }
        "Grep" => {
            let pattern = input.get("pattern").and_then(|v| v.as_str());
            match pattern {
                Some(p) => format!("Searching for '{}'", truncate(p, 30)),
                None => "Searching files".to_string(),
            }
        }
        "Glob" => {
            let pattern = input.get("pattern").and_then(|v| v.as_str());
            match pattern {
                Some(p) => format!("Finding files: {p}"),
                None => "Finding files".to_string(),
            }
        }
        "Task" => "Running task".to_string(),
        "WebSearch" => {
            let query = input.get("query").and_then(|v| v.as_str());
            match query {
                Some(q) => format!("Searching: {}", truncate(q, 40)),
                None => "Searching the web".to_string(),
            }
        }
        "WebFetch" => {
            let url = input.get("url").and_then(|v| v.as_str());
            match url {
                Some(u) => format!("Fetching {}", short_url(u)),
                None => "Fetching URL".to_string(),
            }
        }
        "TodoWrite" => "Updating tasks".to_string(),
        "TodoRead" => "Reading tasks".to_string(),
        "AskUserQuestion" => "Asking user a question".to_string(),
        "NotebookEdit" => "Editing notebook".to_string(),
        _ => format!("Using {tool}"),
    }
}

/// Detect error patterns in tool output and return a summary.
pub fn summarize_tool_result(_tool: &str, output: &str) -> String {
    if output.is_empty() {
        return "Completed (no output)".to_string();
    }

    let lines: Vec<&str> = output.lines().collect();

    // Rust compiler error
    for line in &lines {
        if line.contains("error[E") {
            return format!("Failed: {}", truncate(line.trim(), 60));
        }
    }

    // Generic Error:
    for line in &lines {
        if regex_matches_error(line) {
            return format!("Failed: {}", truncate(line.trim(), 60));
        }
    }

    // FAILED pattern
    if output.contains("FAILED") {
        return "Tests failed".to_string();
    }

    // Panic
    for line in &lines {
        if line.contains("panic!") || line.contains("panicked at") {
            return format!("Panicked: {}", truncate(line.trim(), 60));
        }
    }

    // npm error
    if output.contains("npm ERR!") {
        return "npm error".to_string();
    }

    format!("Completed ({} lines of output)", lines.len())
}

fn regex_matches_error(line: &str) -> bool {
    // Case-insensitive match for word boundary + "Error:"
    let lower = line.to_lowercase();
    // Find "error:" with a word boundary before it
    if let Some(idx) = lower.find("error:") {
        if idx == 0 {
            return true;
        }
        let prev = lower.as_bytes()[idx - 1];
        // Word boundary: not alphanumeric
        return !prev.is_ascii_alphanumeric();
    }
    false
}
